### PR TITLE
[clang-tidy] Add check hicpp-ignored-remove-result

### PR DESCRIPTION
**Pull request by bjosv**
This check implements the [rule 17.5.1](https://www.perforce.com/resources/qac/high-integrity-cpp-coding-standard/standard-library) of the HICPP standard which states:

- Do not ignore the result of std::remove, std::remove_if or std::unique

  The mutating algorithms std::remove, std::remove_if and both overloads of std::unique operate by swapping or moving elements of the range they are operating over. On completion, they return an iterator to the last valid element. In the majority of cases the correct behavior is to use this result as the first operand in a call to std::erase.

This check is based on `bugprone-unused-return-value` but with a fixed set of functions.

Suppressing issues by casting to `void` is enabled by default, but can be disabled by setting `AllowCastToVoid` option to `false`.

### CODE REVIEW

File: clang-tools-extra/clang-tidy/hicpp/HICPPTidyModule.cpp [hunk 1]
```diff
@@ -41,6 +42,15 @@
 #include "NoAssemblerCheck.h"
 #include "SignedBitwiseCheck.h"
 
+namespace {
+
+// Checked functions for hicpp-ignored-remove-result.
+const llvm::StringRef CheckedFunctions = "::std::remove;"
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/hicpp/HICPPTidyModule.cpp:48]: mvoe this to getModuleOptions

File: clang-tools-extra/docs/clang-tidy/checks/list.rst [hunk 2]
```diff
@@ -226,6 +226,7 @@ Clang-Tidy Checks
    :doc:`google-runtime-operator <google/runtime-operator>`,
    :doc:`google-upgrade-googletest-case <google/upgrade-googletest-case>`, "Yes"
    :doc:`hicpp-exception-baseclass <hicpp/exception-baseclass>`,
+   :doc:`hicpp-ignored-remove-result <hicpp/ignored-remove-result>`,
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/list.rst:229]: you added this as new check, not as an alias, aliases are listed bellow, that's why this is confusing.

**Review by PiotrZSL — COMMENTED** [review]
Looks fine, as we got already such checks that are and are not aliases.
But consider doing this as an actually separate check instead of using an alias functionality.
Simply change original check, add other constructor, split storeOptions method, so in this check list of functions would not be configurable by end user.

**Comment by bjosv** [comment]
Great, I'll have a go at it to move it into a separate check.
I used `cert-err33-c` as a blueprint and maybe that check can be updated later as well if we find a good model.

**Comment by PiotrZSL** [comment]
by separate check I mean "Create class that derive from oryginal check, move some code in oryginal check  (config) to protected methods (mainly storeConfig) and just override config in parent class. SImply if you call --dump-config, then for your check list of functions shouldn't be configurable.

**Comment by bjosv** [comment]
Updated the check to be derived from the original check instead of an alias.

File: clang-tools-extra/clang-tidy/bugprone/UnusedReturnValueCheck.cpp [hunk 3]
```diff
@@ -133,6 +133,11 @@ UnusedReturnValueCheck::UnusedReturnValueCheck(llvm::StringRef Name,
                                             "::boost::system::error_code"))),
       AllowCastToVoid(Options.get("AllowCastToVoid", false)) {}
 
+UnusedReturnValueCheck::UnusedReturnValueCheck(llvm::StringRef Name,
+                                               ClangTidyContext *Context,
+                                               std::string CheckedFunctions)
+    : ClangTidyCheck(Name, Context), CheckedFunctions(CheckedFunctions) {}
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/UnusedReturnValueCheck.cpp:139]: std::move(CheckedFunctions)

File: clang-tools-extra/clang-tidy/bugprone/UnusedReturnValueCheck.cpp [hunk 4]
```diff
@@ -133,6 +133,11 @@ UnusedReturnValueCheck::UnusedReturnValueCheck(llvm::StringRef Name,
                                             "::boost::system::error_code"))),
       AllowCastToVoid(Options.get("AllowCastToVoid", false)) {}
 
+UnusedReturnValueCheck::UnusedReturnValueCheck(llvm::StringRef Name,
```
[the remark this answers is not in the document]
Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/bugprone/UnusedReturnValueCheck.cpp:136]: or constructor argument.

**Review by PiotrZSL — COMMENTED** [review]
Small nits, but looks fine.

File: clang-tools-extra/clang-tidy/hicpp/IgnoredRemoveResultCheck.cpp [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+#include "IgnoredRemoveResultCheck.h"
+
+namespace clang::tidy::hicpp {
+
+IgnoredRemoveResultCheck::IgnoredRemoveResultCheck(llvm::StringRef Name,
+                                                   ClangTidyContext *Context)
+    : UnusedReturnValueCheck(Name, Context,
+                             "::std::remove;"
+                             "::std::remove_if;"
+                             "::std::unique;") {
```
[the remark this answers is not in the document]
Comment (bjosv) [reply clang-tools-extra/clang-tidy/hicpp/IgnoredRemoveResultCheck.cpp]: Removed

File: clang-tools-extra/docs/clang-tidy/checks/hicpp/ignored-remove-result.rst [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+`rule 17.5.1 <https://www.perforce.com/resources/qac/high-integrity-cpp-coding-standard/standard-library>`_.
+
+The mutating algorithms ``std::remove``, ``std::remove_if`` and both overloads
+of ``std::unique`` operate by swapping or moving elements of the range they are
+operating over. On completion, they return an iterator to the last valid
+element. In the majority of cases the correct behavior is to use this result as
+the first operand in a call to ``std::erase``.
+
+Suppressing issues by casting to ``void`` is enabled by default and can be
+disabled by setting `AllowCastToVoid` option to ``false``.
```
[the remark this answers is not in the document]
Comment (bjosv) [reply clang-tools-extra/docs/clang-tidy/checks/hicpp/ignored-remove-result.rst]: Options part copied from unused-return-value.rst

File: clang-tools-extra/clang-tidy/bugprone/UnusedReturnValueCheck.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (bjosv) [reply clang-tools-extra/clang-tidy/bugprone/UnusedReturnValueCheck.cpp:136]: Added a constructor that initialized the (missing) CheckedReturnTypes and AllowCastToVoid, which is delegated to from the constructor that is used by the new checker.
Is this in line with your thoughts?

An alternative is to only have the constructor which sets all members, but I thought that looked a bit weird to set AllowCastToVoid twice when constructing IgnoredRemoveResultCheck, once when initializing the baseclass and secondly in the body to be able to use Options.get (which can't be called in the initialization list there).

File: clang-tools-extra/docs/clang-tidy/checks/hicpp/ignored-remove-result.rst [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+Ensure that the result of ``std::remove``, ``std::remove_if`` and ``std::unique``
+are not ignored according to
+`rule 17.5.1 <https://www.perforce.com/resources/qac/high-integrity-cpp-coding-standard/standard-library>`_.
+
+The mutating algorithms ``std::remove``, ``std::remove_if`` and both overloads
+of ``std::unique`` operate by swapping or moving elements of the range they are
+operating over. On completion, they return an iterator to the last valid
+element. In the majority of cases the correct behavior is to use this result as
+the first operand in a call to ``std::erase``.
+
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/hicpp/ignored-remove-result.rst:15]: Add info that this check is a subset of bugprone-unused-return-value, and if that's enabled then this doesn't have to be.

**Review by PiotrZSL — APPROVED** [review]
LGTM

**Pull request merged**